content/blog/managing-kubernetes-infrastructure-with-dotnet-and-pulumi/index.md

```diff
@@ -0,0 +1,251 @@
+
+For example, we can deploy a managed Kubernetes cluster on GKE, and then deploy a Pod into it:
+
+```csharp
+var cluster = new KubernetesCluster("do-cluser", new KubernetesClusterArgs {
+    Region = "sfo2",
+    Version = "latest",
+    NodePool = new KubernetesClusterNodePoolArgs {
+        Name = "default",
+        Size = "s-2vcpu-2gb",
+        NodeCount = nodeCount
+    },
+});
+
+var k8sProvider = new Pulumi.Kubernetes.Provider("do-k8s", new Pulumi.Kubernetes.ProviderArgs {
+    KubeConfig = cluster.KubeConfigs.Apply(array => array[0].RawConfig)
+});
+
+var app = new Pulumi.Kubernetes.Apps.V1.Deployment("do-app-dep", new DeploymentArgs {
+    Spec = new DeploymentSpecArgs {
+        Selector = new LabelSelectorArgs {
+            MatchLabels = new InputMap<string>() {
+                {"app", "app-nginx"},
+            },
+        },
+        Replicas = appReplicaCount,
+        Template = new PodTemplateSpecArgs {
+            Metadata = new ObjectMetaArgs() {
+                Labels = new InputMap<string>() {
+                    {"app", "app-nginx"},
+                },
+            },
+            Spec = new PodSpecArgs() {
+                Containers = new ContainerArgs() {
+                    Name = "nginx",
+                    Image = "nginx",
+                }
+            }
+        }
+    },
+}, new CustomResourceOptions {
+    Provider = k8sProvider
+});
+
+var appService = new Pulumi.Kubernetes.Core.V1.Service("do-app-svc", new ServiceArgs {
+    Spec = new ServiceSpecArgs() {
+        Type = "LoadBalancer",
+        Selector = app.Spec.Apply(spec => spec.Template.Metadata.Labels),
+        Ports = new ServicePortArgs {
+            Port = 80,
+        }
+    }
+}, new CustomResourceOptions {
+    Provider = k8sProvider
+});
+
+var ingressIp = appService.Status.Apply(status => status.LoadBalancer.Ingress[0].Ip);
+
+if (!string.IsNullOrWhiteSpace(domainName)) {
+    var domain = new Domain("do-domain", new DomainArgs() {
+        Name = domainName,
+        IpAddress = ingressIp,
+    });
+
+    var cnameRecord = new DnsRecord("do-domain-cname", new DnsRecordArgs {
+        Domain = domain.Name,
+        Type = "CNAME",
+        Name = "www",
+        Value = "@",
+    });
+}
+```
+
+Note that this example deploys resources first into DigitalOcean (a Kubernetes `Cluster`), then into Kubernetes itself (`Deployment` and `Service` via a `Pulumi.Kubernetes.Provider` configured to connect to the Kuberentes cluster in Digital Ocean), then optionally also more resources in DigitalOcean dependent on the Kubernetes `Service` (`Domain` and `DnsRecord`).  This is a fairly complex infrastructure deployment orchestration, all in a few dozen lines of declarative and strongly typed C# code. 
```

Comment:
Not sure how many .NET devs have heard of DO... Is it a lot more code to make it AKS/EKS?

Reply:
Unfortunately yes - due to the mess around reconstructing `kubeconfig`.  We could possibly just hide that behind a function that we don't include in the example though.  I'll see if we can do that.  

I'll admit though - I kind of liked that this example is (a) quite simple (b) shows some of the diversity of cloud support incidentally and (c) shows cloud resources dependent on Kubernetes resources.

Reply:
After looking again - going to stick with Digital Ocean for this post.